homeassistant/components/goodwe/button.py

```diff
@@ -0,0 +1,59 @@
+"""GoodWe PV inverter selection settings entities."""
+from datetime import datetime
+import logging
+
+from goodwe import Inverter, InverterError
+
+from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
+from homeassistant.helpers.entity import DeviceInfo, EntityCategory
+
+from .const import DOMAIN, KEY_DEVICE_INFO, KEY_INVERTER
+
+_LOGGER = logging.getLogger(__name__)
+
+
+SYNCHRONIZE_CLOCK = ButtonEntityDescription(
+    key="synchronize_clock",
+    name="Synchronize inverter clock",
+    icon="mdi:clock-check-outline",
+    entity_category=EntityCategory.CONFIG,
+)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the inverter select entities from a config entry."""
+    inverter = hass.data[DOMAIN][config_entry.entry_id][KEY_INVERTER]
+    device_info = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE_INFO]
+
+    # read current time from the inverter
+    try:
+        await inverter.read_setting("time")
+    except (InverterError, ValueError):
+        # Inverter model does not support clock synchronization
+        _LOGGER.debug("Could not read inverter current clock time")
+    else:
+        async_add_entities(
+            [SynchronizeInverterClockEntity(device_info, SYNCHRONIZE_CLOCK, inverter)]
+        )
+
+
+class SynchronizeInverterClockEntity(ButtonEntity):
+    """Entity representing the inverter clock synchronization button."""
+
+    _attr_should_poll = False
+
+    def __init__(
+        self,
+        device_info: DeviceInfo,
+        description: ButtonEntityDescription,
+        inverter: Inverter,
+    ) -> None:
+        """Initialize the inverter operation mode setting entity."""
+        self.entity_description = description
+        self._attr_unique_id = f"{DOMAIN}-{description.key}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._inverter: Inverter = inverter
+
+    async def async_press(self) -> None:
+        """Send out a persistent notification."""
+        await self._inverter.write_setting("time", datetime.now())
```

Comment:
It would probably be a good idea to use a lambda function or just function refrence for this and move the specific call to the ButtonEntityDescription.
That will make it a lot esier to add aditional button entities in the future (withouth having to create seperate classes for all of them).

Reply:
@mletenay for example look at the Netgear integration:

https://github.com/home-assistant/core/blob/184421dae68c8f97d7a9f796e2a26137089de562/homeassistant/components/netgear/button.py#L35-L43

https://github.com/home-assistant/core/blob/184421dae68c8f97d7a9f796e2a26137089de562/homeassistant/components/netgear/button.py#L75-L78

Reply:
Good point, done.